poc orch docs, key-concepts- introduction, and dapp pages (#1132)
Agoric/documentation: main/guides/orchestration/getting-started/key-concepts.md

# Orchestration Key Concepts

> Here, we overview the fundamental concepts involved with building orchestration smart contracts.

Reviewer: ```suggestion
This section provides an overview of the fundamental concepts involved in building orchestration smart contracts.
```

### Interchain Account (ICA)

[Interchain Accounts](/glossary/#interchain-account-ica) (ICAs) are an IBC feature used in Agoric’s orchestration API. They enable an Agoric smart contract to control an account on another blockchain within the Cosmos ecosystem, facilitated by Agoric [Orchestration](#orchestration) API. This feature leverages the [Inter-Blockchain Communication (IBC)](#ibc) protocol to facilitate interactions and transactions across different blockchains seamlessly.

<br/>
<img src="../assets/icaoverview.png" width="100%" />
<br/>

Photo credit: [cosmos.network documentation](https://tutorials.cosmos.network/academy/3-ibc/8-ica.html)

> A key advantage of ICAs is that they make accounts on other chains look like any other (remotable) object, meaning normal object access rules (introduction/parenthood/endowment/initial conditions) apply. When a contract creates an ICA, it has sole access to and control over the account but can delegate certain forms of access to its clients. 

Reviewer: ```suggestion
A key advantage of ICAs is that they make accounts on other chains appear like any other (remotable) object, meaning normal object access rules (introduction/parenthood/endowment/initial conditions) apply. When a contract creates an ICA, it has sole access to and control over the account but can delegate certain access permissions to its clients. 
```

For a detailed explanation of these access control rules, see [Access Control with Objects](/guides/zoe/contract-access-control).

### Example ICA Usage from a Smart Contract
This sample is taken from one of the [example contracts](https://github.com/Agoric/agoric-sdk/blob/master/packages/orchestration/src/examples/swapExample.contract.js)

```javascript
const stackAndSwapFn = async (orch, ...) => {
...
  const omni = await orch.getChain('omniflixhub');
  const agoric = await orch.getChain('agoric');

  const [omniAccount, localAccount] = await Promise.all([
    omni.makeAccount(),
    agoric.makeAccount(),
  ]);

  const omniAddress = omniAccount.getAddress();

  // deposit funds from user seat to LocalChainAccount
...
  const transferMsg = orcUtils.makeOsmosisSwap({ ... });

  await localAccount
    .transferSteps(give.Stable, transferMsg)
    .then(_txResult =>
      omniAccount.delegate(offerArgs.validator, offerArgs.staked),
    )
    .catch(e => console.error(e));
};
```

### ChainHub

> The `makeChainHub` utility manages the connections and metadata for various blockchain networks. It simplifies accessing and interacting with multiple chains, providing a unified interface for the orchestration logic to manage cross-chain operations effectively. ChainHub also allows for dynamic registration and usage of chains.

Reviewer: ```suggestion
The `makeChainHub` utility manages the connections and metadata for various blockchain networks. It simplifies access and interaction with multiple chains, providing a unified interface for the orchestration logic to manage cross-chain operations effectively. ChainHub also allows for dynamic registration and usage of chains.
```

1. 	Address Management
```javascript
const address = await orchestrationAccount.getAddress();
```

2.	Balance Management
- getBalances returns an array of amounts for every balance in the account.
-	getBalance retrieves the balance of a specific denom for the account.
```javascript
const balances = await orchestrationAccount.getBalances();
const balance = await orchestrationAccount.getBalance('uatom');
```

3.	Funds Transfer
- send transfers an amount to another account on the same chain.
- transfer transfers an amount to another account, typically on another chain.
- transferSteps transfers an amount in multiple steps, handling complex transfer paths.

```javascript
await orchestrationAccount.send(receiverAddress, amount);
await orchestrationAccount.transfer(amount, destinationAddress);
await orchestrationAccount.transferSteps(amount, transferMsg);
```

> To see the function the Orchestration API exposes, see [Orchestration API](/guides/orchestration/getting-started/api.html)

Reviewer: ```suggestion
For more information on the functions provided, refer to the [Orchestration API](/guides/orchestration/getting-started/api.html) documentation.
```

This composability allows for the development of user-centric applications
 that leverage the unique strengths of different blockchain ecosystems.

The Agoric Orchestration API simplifies interactions between multiple networks, particularly those
 using the [Inter-Blockchain Communication (IBC)](/glossary/#ibc) protocol within Cosmos. The API acts as an 
 abstraction layer, streamlining multi-step processes.

Orchestration integrates with existing Agoric components ([SwingSet](/guides/platform/#swingset), Cosmos modules) and introduces
 vat-orchestration. This [vat](/glossary/#vat) manages Inter-Chain Account (ICA) identities and connections to host
  chains, ensuring proper transaction authorization. 

The orchestration client handles asynchronous tasks and complex workflows, including those 
spanning multiple chains. This empowers smart contracts for actions like inter-chain staking and 
multi-hop transfers, facilitated by notifications from the transfer vat and IBC middleware 
updates. Orchestration simplifies complex cross-chain interactions within a secure and 

> user-controlled environment on the Agoric platform.

Author: left it from moving a section

<!-- ```javascript
lorem
```  -->

## Understanding Key Concepts
Here, we overview the fundamental concepts involved with building orchestration smart contracts.

> ### Interchain Account (ICA)

Author: done

[Interchain Accounts](https://tutorials.cosmos.network/academy/3-ibc/8-ica.html) (ICAs) are an IBC feature used in Agoric’s orchestration API. They enable an Agoric smart contract to control an account on another blockchain within the Cosmos ecosystem, facilitated by Agoric [Orchestration](#orchestration) API. This feature leverages the [Inter-Blockchain Communication (IBC)](#ibc) protocol to facilitate interactions and transactions across different blockchains seamlessly.

<br/>

> <img src="../assets/icaoverview.png" width="100%" />

Author: yea source of image goes into the packet flow details:
https://tutorials.cosmos.network/academy/3-ibc/8-ica.html

adding it

> <img src="../assets/icaoverview.png" width="100%" />

Author: added source, for commit message are you suggesting put the LICENSE its under?

> <img src="../assets/icaoverview.png" width="100%" />

Reviewer 2: yes, a LICENSE would do it

          seat.exit();
          const holder = await makeAccountKit();
          return holder.asContinuingOffer();
        },
        'wantStakingAccount',
      );
    },
  },
);
```

### Zones
Each [Zone](/glossary/#zone) provides an API that allows the allocation of [Exo objects](#exo) and Stores [(object collections)](https://github.com/Agoric/agoric-sdk/tree/master/packages/store/README.md) which use the same underlying persistence mechanism. This allows library code to be agnostic to whether its objects are backed purely by the JS heap (ephemeral), pageable out to disk (virtual), or can be revived after a vat upgrade (durable).

See [SwingSet vat upgrade documentation](https://github.com/Agoric/agoric-sdk/tree/master/packages/SwingSet/docs/vat-upgrade.md) for more example use of the zone API.

```javascript
const zone = makeDurableZone(baggage);
...
zone.subZone('vows')
```

### ChainHub

The `makeChainHub` utility manages the connections and metadata for various blockchain networks. It simplifies accessing and interacting with multiple chains, providing a unified interface for the orchestration logic to manage cross-chain operations effectively.

```javascript

> const chainHub = makeChainHub(remotePowers.agoricNames);

Reviewer 2: suggestion: include something in the code snippet that shows why **ChainHub** is not just a detail we could/should hide altogether.

For example, it lets you register and use chains before the BLD stakers have given them a convenient name:

```js
        chainHub.registerChain(chainKey, chainInfo);
        chainHub.registerConnection(
          agoricChainInfo.chainId,
          chainInfo.chainId,
          connectionInfo,
        );
```

https://github.com/Agoric/agoric-sdk/blob/6bc363b5624bab5fd151ec4889b5f5116f2cf53c/packages/orchestration/src/examples/sendAnywhere.contract.js#L159-L164

Also: I was going to refer you to `ChainHub` in [`@agoric/orchestration` reference docs](https://agoric-sdk.pages.dev/modules/_agoric_orchestration), but it doesn't seem to be there. That's a bug. IOU an issue about fixing those generated reference docs.

Author: added update

```

> # Recorder Kit Makers

Reviewer 2: **technical**: Recorders are not an orchestration concept. They're for vstorage.

These are used to record state changes and events within the contract. `prepareRecorderKitMakers` is a utility that prepares the tools necessary to create these kits, ensuring that the contract can maintain a detailed log of its operations for auditing and debugging purposes.

```javascript
const { makeRecorderKit } = prepareRecorderKitMakers(baggage, marshaller);
const recorderKit = makeRecorderKit(
    privateArgs.storageNode,
    /** @type {import('@agoric/zoe/src/contractSupport/recorder.js').TypedMatcher<MetricsNotification>} */ (
      M.any()
    ),
)
await recorderKit.recorder.write(proposalsArray);
```

> ### Vow Tools

Reviewer 2: **technical**: Vows are not an orchestration concept. They're for dealing with promises across upgrade.

**editorial**: Perhaps rather than "Orchestration Concept", this page should be "Orchestration contract walkthru", as a follow-up to the [contract walkthru that demonstrates trading with offer safety](https://docs.agoric.com/guides/zoe/contract-walkthru.html). These sections would be verb phrases: **Using Vows to resume after upgrade** and such.

Author: added vow to glossary, and moved `vow tools` to contract upgrade. writing a contract walkthrough

These tools handle promises and asynchronous operations within the contract. `prepareVowTools` prepares the necessary utilities to manage these asynchronous tasks, ensuring that the contract can handle complex workflows that involve waiting for events or responses from other chains.

```javascript
const vowTools = prepareVowTools(zone.subZone('vows'));
...
const makeLocalOrchestrationAccountKit = prepareLocalChainAccountKit(
  zone,
  makeRecorderKit,
  zcf,
  privateArgs.timerService,
  vowTools,
  makeChainHub(privateArgs.agoricNames),
);
...
const makeCosmosOrchestrationAccount = prepareCosmosOrchestrationAccount(
  zone,
  makeRecorderKit,
  vowTools,
  zcf,
);
```

> ### Durable Zone

Reviewer 2: not an orchestration concept; Zones are for upgrade

A zone specifically designed for durability, allowing the contract to persist its state across upgrades. This is critical for maintaining the continuity and reliability of the contract’s operations.
```javascript
const zone = makeDurableZone(baggage);
```

> ### Account Kits

Reviewer 2: **Orchestration Account** is a key concept, represented as `OrchestrationAccountI` in [orchestration-api.ts](https://github.com/Agoric/agoric-sdk/blob/master/packages/orchestration/src/orchestration-api.ts).

"Kit" doesn't belong in the name of a key concept. An "X Kit" just means "An X and some related facets/stuff", as explained in the section on [Method Naming Structure](https://docs.agoric.com/guides/ertp/#method-naming-structure) _which is in an ERTP guide for no reason; it should probably go in the [JavaScript framework section](https://docs.agoric.com/guides/js-programming/)... probably under [separation of duties](https://docs.agoric.com/guides/js-programming/hardened-js.html#separation-of-duties)... and we should introduce the term **facet** near there_.
